src/libs/ajax/Groups.ts

```diff
@@ -0,0 +1,89 @@
+import _ from 'lodash/fp';
+import { authOpts, fetchSam, jsonBody } from 'src/libs/ajax/ajax-common';
+
+export type GroupRole = 'admin' | 'member';
+
+export interface CurrentUserGroupMembership {
+  groupEmail: string;
+  groupName: string;
+  role: GroupRole;
+}
+
+export const Groups = (signal?: AbortSignal) => ({
+  list: async (): Promise<CurrentUserGroupMembership[]> => {
+    const res = await fetchSam('api/groups/v1', _.merge(authOpts(), { signal }));
+    return res.json();
+  },
+
+  group: (groupName: string) => {
+    const root = `api/groups/v1/${groupName}`;
+    const resourceRoot = `api/resources/v1/managed-group/${groupName}`;
+
+    const addRole = (role: GroupRole, email: string): Promise<void> => {
+      return fetchSam(`${root}/${role}/${encodeURIComponent(email)}`, _.merge(authOpts(), { signal, method: 'PUT' }));
+    };
+
+    const removeRole = (role: GroupRole, email: string): Promise<void> => {
+      return fetchSam(
+        `${root}/${role}/${encodeURIComponent(email)}`,
+        _.merge(authOpts(), { signal, method: 'DELETE' })
+      );
+    };
+
+    return {
+      create: (): Promise<void> => {
+        return fetchSam(root, _.merge(authOpts(), { signal, method: 'POST' }));
+      },
+
+      delete: (): Promise<void> => {
+        return fetchSam(root, _.merge(authOpts(), { signal, method: 'DELETE' }));
+      },
+
+      listAdmins: async (): Promise<string[]> => {
+        const res = await fetchSam(`${root}/admin`, _.merge(authOpts(), { signal }));
+        return res.json();
+      },
+
+      listMembers: async (): Promise<string> => {
+        const res = await fetchSam(`${root}/member`, _.merge(authOpts(), { signal }));
+        return res.json();
+      },
+
+      addUser: (roles: GroupRole[], email: string): Promise<void[]> => {
+        return Promise.all(_.map((role) => addRole(role, email), roles));
+      },
+
+      removeUser: (roles: GroupRole[], email: string): Promise<void[]> => {
+        return Promise.all(_.map((role) => removeRole(role, email), roles));
+      },
+
+      changeUserRoles: async (email: string, oldRoles: GroupRole[], newRoles: GroupRole[]): Promise<void[] | void> => {
+        if (!_.isEqual(oldRoles, newRoles)) {
+          await Promise.all(_.map((role: GroupRole) => addRole(role, email), _.difference(newRoles, oldRoles)));
+          return Promise.all(_.map((role: GroupRole) => removeRole(role, email), _.difference(oldRoles, newRoles)));
+        }
+      },
+
+      requestAccess: async (): Promise<void> => {
+        await fetchSam(`${root}/requestAccess`, _.merge(authOpts(), { signal, method: 'POST' }));
+      },
+
+      getPolicy: async (policyName: string): Promise<boolean> => {
```

Comment:
I would not have expected that `getPolicy` returns a boolean! But I checked the code… and sure enough.

Reply:
Yeah, I initially typed this as `any` because I thought it was getting some policy configuration. Then I realized it was just a boolean. This method definitely seems poorly named.